File: Python/Observer.py
class ISubject:
    def addWatch(self, IWatchers):
        pass
    def notify(self):
        pass
class IWatchers:
    def alert(self):
        pass

class Alarm(ISubject):
    def __init__(self):
        self.__watchers = []
    def addWatch(self, IWatchers):
        self.__watchers.append(IWatchers)
    def notify(self):
        for watcher in self.__watchers:
            watcher.alert()

class FireStation(IWatchers):
    def alert(self):
        print("Bombeiros respondendo! Estamos a caminho")

class PoliceStation(IWatchers):
    def alert(self):
        print("Policia respondendo! Estamos a caminho")

class HospitalStation(IWatchers):
    def alert(self):
        print("Hospital respondendo! Estamos a caminho")

class Main:
    def main(self):
        alarm = Alarm()
        alarm.addWatch(FireStation())
        alarm.addWatch(HospitalStation())
        alarm.addWatch(PoliceStation())
        alarm.notify()

if __name__ == "__main__":
    Main().main()

File: Ruby/observer.rb
#!/bin/ruby

class ISubject
  def add_watch watchers
  end
  def notify
  end
end

class IWatchers
  def alert
  end
end

class Alarm < ISubject
  def initialize
    @watchers = Array.new
  end
  def add_watch watchers
    @watchers << watchers
  end
  def notify
    @watchers.each do |watcher|
      watcher.alert()
    end
  end
end

class FireStation < IWatchers
  def alert
    puts "Bombeiros respondendo! Estamos a caminho"
  end
end

class PoliceStation < IWatchers
  def alert
    puts "Policia respondendo! Estamos a caminho"
  end
end

class HospitalStation < IWatchers
  def alert
    puts "Hospital respondendo! Estamos a caminho"
  end
end

alarm = Alarm.new
alarm.add_watch FireStation.new
alarm.add_watch HospitalStation.new
alarm.add_watch PoliceStation.new
alarm.notify
